test/parallel/test-http2-client-rststream-before-connect.js

```diff
@@ -15,12 +15,15 @@ server.on('listening', common.mustCall(() => {
   const client = h2.connect(`http://localhost:${server.address().port}`);
 
   const req = client.request({ ':path': '/' });
+  // make sure that destroy is called twice
+  req._destroy = common.mustCall(req._destroy.bind(req), 2);
```

Comment:
I am a bit lost in how this is possible, in theory it is protected from running `_destroy` twice: https://github.com/nodejs/node/blob/master/lib/internal/streams/destroy.js#L10-L30

Reply:
It was confusing for me too.
The two instances where `stream.destroy()` is called is:
- in first call of rstStream() https://github.com/nodejs/node/blob/ff747e3fe8fc960bb6cd04185167bd6376940998/lib/internal/http2/core.js#L631
- in second call of rstStream() https://github.com/nodejs/node/blob/ff747e3fe8fc960bb6cd04185167bd6376940998/lib/internal/http2/core.js#L931

This might translate to two calls of `req._destroy`?

I've run the test on repeat 100 times to confirm that it it succeeds using the following command:
```
> python tools/test.py -J --mode=release parallel/test-http2-client-rststream-before-connect --repeat 100
[00:06|% 100|+ 100|-   0]: Done
```

Reply:
I believe @mcollina is right. I think what's happening is that this line is accounting for the 2 triggers of _destroy:

https://github.com/nodejs/node/blob/ff747e3fe8fc960bb6cd04185167bd6376940998/lib/internal/http2/core.js#L1494-L1498

The reason that it's different than before is because the declaration was moved before `client.rstStream()` instead of after. I think that `stream.destroy()` isn't even necessary and it should just be the empty `return;`.

ping @jasnell — any thoughts re: whether that `stream.destroy()` serves an actual purpose?

Reply:
> I've run the test on repeat 100 times

@trivikr the test might be passing, but I'm wondering if the actual behavior is correct.

I don't think that calling `rstStream()` twice should destroy the stream. It seems a not needed side effect. I think it should throw. https://github.com/nodejs/node/blob/ff747e3fe8fc960bb6cd04185167bd6376940998/lib/internal/http2/core.js#L929-L931

cc @jasnell

Reply:
@mcollina I ran the test on repeat 100 times to check if there's any flakiness in it.
I'll wait for the response from @jasnell and update the code and test as per the conclusion.

Reply:
I'll dig in on this a bit more later on today